sw/vendor/veri-titan/gen/otbn_modexp.s

```diff
@@ -1,133 +1,348 @@
+   *
+   * The squared Montgomery modulus RR and the Montgomery constant m0' have to
+   * be provided at the appropriate locations in dmem. DMEM locations are
+   * expected to be disjoint.
+   *
+   * Flags: Flags have no meaning beyond the scope of this subroutine.
+   *
+   * The base bignum A is expected in the input buffer, the result C is written
+   * to the output buffer.
+   *
+   * @param[in]  dmem[x17] pointer to m0' in dmem
+   * @param[in]  dmem[x26] pointer to RR in dmem
+   * @param[in]  dmem[x16] pointer to first limb of modulus M in dmem
+   * @param[in]  dmem[x23] pointer to buffer with base bignum
+   * @param[in]  dmem[x24] pointer to output buffer
+   */
+
+  /* Prepare all-zero register. */
   bn.xor w31, w31, w31 << 0, FG0
+
+  /* Prepare pointers to temporary registers. */
   li x8, 4
   li x9, 3
   li x10, 4
   li x11, 2
+
+  /* Initialize a counter to double-check that the loop completes (a
+      protection against fault injection). */
+  li x5, 0
+
+  /* Convert input to Montgomery domain.
+       [w15:w4] <= (dmem[x23]*dmem[x26]*R^-1) mod M = (A * R) mod M */
   addi x19, x23, 0
   addi x20, x26, 0
   addi x21, x24, 0
   jal x1, montmul
+
+  /* Store Montgomery-form input in dmem.
+       dmem[x24] <= [w15:w4] = (A * R) mod M */
   loopi 12, 3
     bn.sid x8, 0(x21++)
     addi x8, x8, 1
-    nop
+    addi x5, x5, 1
+
+  /* 16 consecutive Montgomery squares on the outbut buffer, i.e. after loop:
+       dmem[out_buf] <= dmem[out_buf]^65536*R mod M */
   loopi 16, 9
+
+    /* dmem[out_buf]  <= montmul(dmem[out_buf], dmem[out_buf]) */
     addi x19, x24, 0
     addi x20, x24, 0
     addi x21, x24, 0
     jal x1, montmul
     loopi 12, 3
       bn.sid x8, 0(x21++)
       addi x8, x8, 1
-      nop
-    nop
+      addi x5, x5, 1
+
+    /* Update counter. */
+    addi x5, x5, 1
+
+  /* Final multiplication and conversion of result from Montgomery domain.
+       out_buf  <= montmul(*x28, *x20) = montmul(dmem[in_buf], dmem[out_buf]) */
   addi x19, x23, 0
   addi x20, x24, 0
   addi x21, x24, 0
   jal x1, montmul
+
+  /* Final conditional subtraction of modulus if mod >= dmem[out_buf]. */
+
+  /* Speculatively subtract modulus and store in separate registers.
+       [w28:w17] <= (dmem[out_buf] - M) mod 2^3072 */
   bn.add w31, w31, w31 << 0, FG0
   li x17, 16
-  loopi 12, 5
+  loopi 12, 4
     bn.movr x11, x8++
     bn.lid x9, 0(x16++)
     bn.subb w2, w2, w3 << 0, FG0
     bn.movr x17++, x11
-    nop
+
+  /* Check the borrow flag and select the post-subtraction version iff the
+     subtraction underflowed. */
   csrrs x2, 1984, x0
   andi x2, x2, 1
   li x8, 4
   bne x2, x0, label_0
   li x8, 16
   label_0:
+
+  /* Store result in dmem: dmem[out_buf] <= A^65537 mod M */
   addi x21, x24, 0
   loopi 12, 3
     bn.sid x8, 0(x21++)
     addi x8, x8, 1
-    nop
+    addi x5, x5, 1
+
+  /* If the counter value doesn't match expectations, cause a deliberate
+     error (WDR reference > 31) to end the program. */
```

Comment:
I realise that this is vendored, so you don't want to change it here. But we have an explicit `unimp` instruction that you can use for this. (It's encoded as a CSR operation on an invalid CSR)

Reply:
Ooh, much better! I'll submit an upstream patch for that and update again once it's merged :slightly_smiling_face: